File: app/src/main/java/com/gnoemes/shikimoriapp/presentation/view/anime/AnimeFragment.java
package com.gnoemes.shikimoriapp.presentation.view.anime;

import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.design.widget.AppBarLayout;
import android.support.design.widget.CoordinatorLayout;
import android.support.transition.Fade;
import android.support.transition.TransitionManager;
import android.support.v4.view.ViewPager;
import android.support.v4.widget.SwipeRefreshLayout;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.arellomobile.mvp.presenter.InjectPresenter;
import com.arellomobile.mvp.presenter.ProvidePresenter;
import com.gnoemes.shikimoriapp.R;
import com.gnoemes.shikimoriapp.entity.anime.domain.AnimeFranchiseNode;
import com.gnoemes.shikimoriapp.entity.anime.presentation.AnimeAction;
import com.gnoemes.shikimoriapp.entity.anime.presentation.AnimeDetailsPage;
import com.gnoemes.shikimoriapp.entity.anime.presentation.AnimeDetailsViewModel;
import com.gnoemes.shikimoriapp.entity.anime.presentation.AnimeLinkViewModel;
import com.gnoemes.shikimoriapp.entity.anime.presentation.delegate.BaseEpisodeItem;
import com.gnoemes.shikimoriapp.entity.anime.series.domain.TranslationType;
import com.gnoemes.shikimoriapp.entity.app.presentation.AppExtras;
import com.gnoemes.shikimoriapp.entity.app.presentation.BaseItem;
import com.gnoemes.shikimoriapp.entity.rates.domain.UserRate;
import com.gnoemes.shikimoriapp.presentation.presenter.anime.AnimePresenter;
import com.gnoemes.shikimoriapp.presentation.presenter.anime.converter.AnimeDetailsViewModelConverter;
import com.gnoemes.shikimoriapp.presentation.view.anime.adapter.anime.AnimeAdapter;
import com.gnoemes.shikimoriapp.presentation.view.anime.adapter.anime.AnimeCharacterAdapter;
import com.gnoemes.shikimoriapp.presentation.view.anime.adapter.comments.CommentsAdapter;
import com.gnoemes.shikimoriapp.presentation.view.anime.adapter.episodes.EpisodeAdapter;
import com.gnoemes.shikimoriapp.presentation.view.anime.converter.AnimeFranchiseNodeToStringConverter;
import com.gnoemes.shikimoriapp.presentation.view.anime.provider.RateResourceProvider;
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.BaseFragment;
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.RouterProvider;
import com.gnoemes.shikimoriapp.utils.Utils;
import com.gnoemes.shikimoriapp.utils.imageloader.ImageLoader;
import com.gnoemes.shikimoriapp.utils.view.AttributesHelper;
import com.gnoemes.shikimoriapp.utils.view.DrawableHelper;
import com.gnoemes.shikimoriapp.utils.view.LinearStickyHead;
import com.gnoemes.shikimoriapp.utils.view.VerticalSpaceItemDecoration;
import com.gnoemes.shikimoriapp.utils.view.ViewStatePagerAdapter;

import net.opacapp.multilinecollapsingtoolbar.CollapsingToolbarLayout;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import javax.inject.Inject;

import butterknife.BindView;

public class AnimeFragment extends BaseFragment<AnimePresenter, AnimeView>
        implements AnimeView {

    @BindView(R.id.coordinator_layout)
    CoordinatorLayout layout;

    @BindView(R.id.app_bar)
    AppBarLayout appBarLayout;

    @BindView(R.id.toolbar)
    Toolbar toolbar;

    @BindView(R.id.collapsing_toolbar)
    CollapsingToolbarLayout collapsingToolbarLayout;

    @BindView(R.id.image_background)
    ImageView backgroundImage;

    @BindView(R.id.view_pager)
    ViewPager viewPager;

    @BindView(R.id.subtitle)
    TextView subtitleView;

    @BindView(R.id.progress_loading)
    ProgressBar progressBar;

    @InjectPresenter
    AnimePresenter presenter;

    @ProvidePresenter
    AnimePresenter provideAnimePresenter() {
        presenter = presenterProvider.get();
        if (getParentFragment() != null) {
            presenter.setLocalRouter(((RouterProvider) getParentFragment()).getLocalRouter());
        }
        if (getArguments() != null) {
            presenter.setAnimeId(getArguments().getLong(AppExtras.ARGUMENT_ANIME_ID));
        }
        return presenter;
    }

    @Inject
    ImageLoader imageLoader;

    @Inject
    AnimeDetailsViewModelConverter converter;

    @Inject
    AnimeFranchiseNodeToStringConverter franchiseConverter;

    @Inject
    RateResourceProvider rateResourceProvider;

    private MaterialDialog translationTypeDialog;
    private AnimePagerAdapter pagerAdapter;
    private boolean isCommentsPage;

    public static AnimeFragment newInstance(long animeId) {
        Bundle args = new Bundle();
        AnimeFragment fragment = new AnimeFragment();
        args.putLong(AppExtras.ARGUMENT_ANIME_ID, animeId);
        fragment.setArguments(args);
        return fragment;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(getFragmentLayout(), container, false);
    }


    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        initViews();
    }

    private ViewPager.OnPageChangeListener pageChangeListener = new ViewPager.SimpleOnPageChangeListener() {
        @Override
        public void onPageSelected(int position) {
            isCommentsPage = AnimeDetailsPage.COMMENTS.isEqualPage(position);
        }
    };
    private AppBarLayout.OnOffsetChangedListener onOffsetChangedListener = new AppBarLayout.OnOffsetChangedListener() {
        @Override
        public void onOffsetChanged(AppBarLayout appBarLayout, int verticalOffset) {
            subtitleView.setAlpha((float) (1 - Math.abs(verticalOffset / 1.5 / 100)));
        }
    };

    @Override
    public void onDestroyView() {
        viewPager.setAdapter(null);
        viewPager.removeOnPageChangeListener(pageChangeListener);
        appBarLayout.removeOnOffsetChangedListener(onOffsetChangedListener);
        backgroundImage.setOnClickListener(null);
        super.onDestroyView();
    }

    private void initViews() {
        AnimeCharacterAdapter characterAdapter = new AnimeCharacterAdapter(imageLoader, id -> getPresenter().onAction(AnimeAction.CHARACTER, id));
        EpisodeAdapter episodeAdapter = new EpisodeAdapter(item -> getPresenter().onEpisodeClicked(item),
                (action, item) -> getPresenter().onEpisodeOptionAction(action, item));
        AnimeAdapter animeAdapter = new AnimeAdapter(rateResourceProvider, characterAdapter, (action, data) -> getPresenter().onAction(action, data));
        CommentsAdapter commentsAdapter = new CommentsAdapter(imageLoader,
                id -> getPresenter().onUserClicked(id));

        pagerAdapter = new AnimePagerAdapter(commentsAdapter, animeAdapter, episodeAdapter);
        viewPager.setAdapter(pagerAdapter);
        viewPager.addOnPageChangeListener(pageChangeListener);

        int textColor = AttributesHelper.withContext(getContext())
                .getColor(R.attr.colorText);

        collapsingToolbarLayout.setCollapsedTitleTextColor(textColor);
        Drawable navigationIcon = DrawableHelper.withContext(getContext())
                .withDrawable(R.drawable.ic_arrow_back)
                .withAttributeColor(R.attr.colorText)
                .tint()
                .get();

        toolbar.setNavigationIcon(navigationIcon);
        toolbar.setNavigationOnClickListener(v -> getPresenter().onBackPressed());

        backgroundImage.setOnClickListener(v -> getPresenter().onBackgroundImageClicked());
        appBarLayout.addOnOffsetChangedListener(onOffsetChangedListener);
    }

    ///////////////////////////////////////////////////////////////////////////
    // GETTERS
    ///////////////////////////////////////////////////////////////////////////

    @Override
    protected AnimePresenter getPresenter() {
        return presenter;
    }

    @Override
    protected int getFragmentLayout() {
        return R.layout.fragment_details_page;
    }

    ///////////////////////////////////////////////////////////////////////////
    // MVP
    ///////////////////////////////////////////////////////////////////////////

    @Override
    public void onShowLoading() {
        TransitionManager.beginDelayedTransition(layout, new Fade());
        progressBar.setVisibility(View.VISIBLE);
        appBarLayout.setVisibility(View.GONE);
        viewPager.setVisibility(View.GONE);
    }

    @Override
    public void onHideLoading() {
        TransitionManager.beginDelayedTransition(layout, new Fade());
        progressBar.setVisibility(View.GONE);
        appBarLayout.setVisibility(View.VISIBLE);
        viewPager.setVisibility(View.VISIBLE);
    }

    @Override
    public void setAnimeData(AnimeDetailsViewModel model) {
        imageLoader.setImageWithFit(backgroundImage, model.getImageUrl());
        collapsingToolbarLayout.setTitle(model.getName());
        subtitleView.setText(model.getJpOrEngName());
        pagerAdapter.setData(converter.convertFromViewModel(model));
    }

    @Override
    public void setPage(int position) {
        viewPager.setCurrentItem(position, true);
    }

    @Override
    public void showEpisodeList(List<BaseEpisodeItem> episodes) {
        pagerAdapter.showEpisodeList(episodes);
    }

    @Override
    public void showComments(List<BaseItem> baseCommentItems) {
        pagerAdapter.showComments(baseCommentItems);
    }

    @Override
    public void insetMoreComments(List<BaseItem> baseCommentItems) {
        pagerAdapter.insertMoreComments(baseCommentItems);
    }

    @Override
    public void onShowPageLoading() {
        pagerAdapter.showPageLoading();
    }

    @Override
    public void onHidePageLoading() {
        pagerAdapter.hidePageLoading();
    }

    @Override
    public void showErrorView() {
        viewPager.setVisibility(View.GONE);
    }

    @Override
    public void hideErrorView() {
        viewPager.setVisibility(View.VISIBLE);
    }

    @Override
    public void onShowRefresh() {
        pagerAdapter.onShowRefresh();
    }

    @Override
    public void onHideRefresh() {
        pagerAdapter.onHideRefresh();
    }

    @Override
    public void showPlayWizard(List<TranslationType> types) {

        if (translationTypeDialog == null || !translationTypeDialog.isShowing()) {
            String[] translationTypes = new String[types.size()];
            for (int i = 0; i < translationTypes.length; i++) {
                translationTypes[i] = Utils.firstUpperCase(types.get(i).getType());
            }

            translationTypeDialog = new MaterialDialog.Builder(getContext())
                    .dividerColor(R.attr.colorAccent)
                    .items(translationTypes)
                    .itemsColorAttr(R.attr.colorText)
                    .backgroundColorAttr(R.attr.colorBackgroundWindow)
                    .buttonRippleColorAttr(R.attr.colorAccentTransparent)
                    .autoDismiss(true)
                    .canceledOnTouchOutside(true)
                    .itemsCallback((dialog, itemView, position, text) -> {
                        getPresenter().onTranslationTypeChoosed(types.get(position));
                    })
                    .build();

            translationTypeDialog.show();
        }
    }

    @Override
    public void showLinksDialog(List<AnimeLinkViewModel> animeLinkViewModels) {

        List<String> titles = new ArrayList<>();

        for (AnimeLinkViewModel model : animeLinkViewModels) {
            titles.add(model.getName());
        }

        new MaterialDialog.Builder(getContext())
                .title(R.string.common_links)
                .items(titles.toArray(new CharSequence[titles.size()]))
                .itemsCallback((dialog, itemView, position, text) -> {
                    dialog.dismiss();
                    AnimeFragment.this.getPresenter().onLinkPressed(animeLinkViewModels.get(position));
                })
                .autoDismiss(true)
                .titleColorAttr(R.attr.colorText)
                .contentColorAttr(R.attr.colorText)
                .alwaysCallSingleChoiceCallback()
                .backgroundColorAttr(R.attr.colorBackgroundWindow)
                .autoDismiss(false)
                .negativeColorAttr(R.attr.colorAction)
                .negativeText(R.string.close)
                .onNegative((dialog, which) -> dialog.dismiss())
                .canceledOnTouchOutside(true)
                .build()
                .show();
    }

    @Override
    public void showChronologyDialog(List<AnimeFranchiseNode> nodes) {

        List<String> items = franchiseConverter.convertList(nodes);

        new MaterialDialog.Builder(getContext())
                .title(R.string.chronology)
                .items(items.toArray(new CharSequence[items.size()]))
                .itemsCallback((dialog, itemView, position, text) -> {
                    dialog.dismiss();
                    if (nodes != null && !nodes.isEmpty()) {
                        getPresenter().onAnimeClicked(nodes.get(position).getId());
                    }
                })
                .autoDismiss(true)
                .titleColorAttr(R.attr.colorText)
                .contentColorAttr(R.attr.colorText)
                .alwaysCallSingleChoiceCallback()
                .backgroundColorAttr(R.attr.colorBackgroundWindow)
                .negativeText(R.string.close)
                .negativeColorAttr(R.attr.colorAction)
                .canceledOnTouchOutside(true)
                .build()
                .show();
    }

    @Override
    public void showRatesDialog(UserRate data) {
        RateDialogFragment dialog = RateDialogFragment.newInstance(data);
        dialog.setCallback(new RateDialogFragment.RateDialogCallback() {
            @Override
            public void onSaveAnimeRate(UserRate rate) {
                getPresenter().onSaveRate(rate);
            }

            @Override
            public void onDeleteAnimeRate(long id) {
                getPresenter().onDeleteRate(id);
            }
        });
        dialog.show(getChildFragmentManager(), "RATES");
    }

    @Override
    public void showClearHistoryDialog() {
        new MaterialDialog.Builder(getContext())
                .content(R.string.clear_episodes_content)
                .autoDismiss(true)
                .titleColorAttr(R.attr.colorText)
                .contentColorAttr(R.attr.colorText)
                .alwaysCallSingleChoiceCallback()
                .backgroundColorAttr(R.attr.colorBackgroundWindow)
                .negativeColorAttr(R.attr.colorAction)
                .negativeText(R.string.common_cancel)
                .onNegative((dialog, which) -> dialog.dismiss())
                .positiveColorAttr(R.attr.colorAction)
                .positiveText(R.string.yes)
                .onPositive((dialog, which) -> getPresenter().onClearHistory())
                .canceledOnTouchOutside(true)
                .build()
                .show();
    }

    ///////////////////////////////////////////////////////////////////////////
    // INNER CLASS
    ///////////////////////////////////////////////////////////////////////////

    public class AnimePagerAdapter extends ViewStatePagerAdapter {


        private final List<Integer> screens = Arrays.asList(
                R.layout.fragment_comments,
                R.layout.fragment_anime,
                R.layout.fragment_series);
        private CommentsAdapter commentsAdapter;
        private AnimeAdapter animeAdapter;
        private EpisodeAdapter episodeAdapter;

        private RecyclerView animeDetailsList;
        private RecyclerView seriesList;
        private RecyclerView commentsList;

        private SwipeRefreshLayout refreshLayout;

        AnimePagerAdapter(CommentsAdapter commentsAdapter, AnimeAdapter animeAdapter, EpisodeAdapter episodeAdapter) {
            this.commentsAdapter = commentsAdapter;
            this.animeAdapter = animeAdapter;
            this.episodeAdapter = episodeAdapter;
        }

        @Override
        public int getCount() {
            return screens.size();
        }

        @Override
        protected View createView(ViewGroup container, int position) {
            LayoutInflater inflater = LayoutInflater.from(container.getContext());
            ViewGroup layout = (ViewGroup) inflater.inflate(screens.get(position), container, false);
            switch (position) {
                case 0:
                    createCommentsPage(layout);
                    break;
                case 1:
                    createAnimePage(layout);
                    break;
                case 2:
                    createSeriesPage(layout);
                    break;
            }
            return layout;
        }

        private void createCommentsPage(ViewGroup layout) {
            commentsList = layout.findViewById(R.id.list_comments);
            LinearLayoutManager layoutManager = new LinearLayoutManager(layout.getContext());
            commentsList.setLayoutManager(layoutManager);
            commentsList.setAdapter(commentsAdapter);
            commentsList.addOnScrollListener(new RecyclerView.OnScrollListener() {
                @Override
                public void onScrolled(RecyclerView recyclerView, int dx, int dy) {
                    super.onScrolled(recyclerView, dx, dy);

                    //TODO fix endless scroll listener (layoutManager.findLastCompletelyVisibleItemPosition returns -1)
                    int visibleItemPosition = layoutManager.findLastVisibleItemPosition();
//                    int visibleItemPosition = layoutManager.findLastCompletelyVisibleItemPosition();
                    int itemCount = layoutManager.getItemCount() - 1;

                    if (visibleItemPosition >= itemCount && isCommentsPage) {
                        getPresenter().loadNextPage();
                    }

                }
            });
        }


        private void createAnimePage(ViewGroup layout) {
            animeDetailsList = layout.findViewById(R.id.anime_details_list);
            animeDetailsList.setItemAnimator(new DefaultItemAnimator());
            animeDetailsList.setAdapter(animeAdapter);
            animeDetailsList.setLayoutManager(new LinearLayoutManager(layout.getContext()));
        }

        private void createSeriesPage(ViewGroup layout) {
            refreshLayout = layout.findViewById(R.id.refresh_layout);
            seriesList = layout.findViewById(R.id.list);
            refreshLayout.setOnRefreshListener(() -> getPresenter().onEpisodesRefresh());
            seriesList.setLayoutManager(new LinearStickyHead<EpisodeAdapter>(layout.getContext()));
            int margin = (int) layout.getResources().getDimension(R.dimen.margin_small);
            seriesList.addItemDecoration(new VerticalSpaceItemDecoration(margin));
            seriesList.setAdapter(episodeAdapter);
            if (episodeAdapter.getItems().isEmpty()) {
                refreshLayout.setRefreshing(true);
            }
        }

        public void setData(List<BaseItem> animeItems) {
            animeAdapter.bindItems(animeItems);
        }

        void showEpisodeList(List<BaseEpisodeItem> episodes) {
            episodeAdapter.bindItems(episodes);
        }


        void onHideRefresh() {
            if (refreshLayout != null) {
                refreshLayout.setRefreshing(false);
            }
        }

        void onShowRefresh() {
            if (refreshLayout != null) {
                refreshLayout.setRefreshing(true);
            }
        }

        void showComments(List<BaseItem> baseCommentItems) {
            commentsAdapter.bindItems(baseCommentItems);
        }

        void insertMoreComments(List<BaseItem> baseCommentItems) {
            commentsAdapter.insertMore(baseCommentItems);
        }

        void hidePageLoading() {
            commentsAdapter.hideProgress();
        }

        void showPageLoading() {
            commentsAdapter.showProgress();
        }
    }
}

File: app/src/main/java/com/gnoemes/shikimoriapp/presentation/view/anime/AnimeFragment.kt
package com.gnoemes.shikimoriapp.presentation.view.anime

import android.os.Bundle
import android.support.design.widget.AppBarLayout
import android.support.transition.Fade
import android.support.transition.TransitionManager
import android.support.v4.view.ViewPager
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.DividerItemDecoration
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.checkbox.checkBoxPrompt
import com.afollestad.materialdialogs.list.listItems
import com.arellomobile.mvp.presenter.InjectPresenter
import com.arellomobile.mvp.presenter.ProvidePresenter
import com.gnoemes.shikimoriapp.R
import com.gnoemes.shikimoriapp.data.local.preferences.UserSettingsSource
import com.gnoemes.shikimoriapp.entity.anime.domain.FranchiseNode
import com.gnoemes.shikimoriapp.entity.anime.presentation.AnimeDetailsPage
import com.gnoemes.shikimoriapp.entity.anime.presentation.AnimeDetailsViewModel
import com.gnoemes.shikimoriapp.entity.anime.presentation.LinkViewModel
import com.gnoemes.shikimoriapp.entity.anime.presentation.delegate.BaseEpisodeItem
import com.gnoemes.shikimoriapp.entity.anime.series.domain.TranslationType
import com.gnoemes.shikimoriapp.entity.app.data.SettingsExtras
import com.gnoemes.shikimoriapp.entity.app.domain.Type
import com.gnoemes.shikimoriapp.entity.app.presentation.AppExtras
import com.gnoemes.shikimoriapp.entity.app.presentation.BaseItem
import com.gnoemes.shikimoriapp.entity.rates.domain.UserRate
import com.gnoemes.shikimoriapp.presentation.presenter.anime.AnimePresenter
import com.gnoemes.shikimoriapp.presentation.presenter.anime.converter.AnimeDetailsViewModelConverter
import com.gnoemes.shikimoriapp.presentation.view.anime.adapter.anime.AnimeAdapter
import com.gnoemes.shikimoriapp.presentation.view.anime.adapter.anime.AnimeItemCallback
import com.gnoemes.shikimoriapp.presentation.view.anime.adapter.anime.DetailsCharacterAdapter
import com.gnoemes.shikimoriapp.presentation.view.anime.adapter.comments.CommentsAdapter
import com.gnoemes.shikimoriapp.presentation.view.anime.adapter.episodes.EpisodeAdapter
import com.gnoemes.shikimoriapp.presentation.view.anime.adapter.episodes.EpisodeOptionCallback
import com.gnoemes.shikimoriapp.presentation.view.anime.adapter.episodes.EpisodePickCallback
import com.gnoemes.shikimoriapp.presentation.view.anime.converter.FranchiseNodeToStringConverterImpl
import com.gnoemes.shikimoriapp.presentation.view.anime.provider.RateResourceProvider
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.BaseFragment
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.RateDialogFragment
import com.gnoemes.shikimoriapp.presentation.view.common.fragment.RouterProvider
import com.gnoemes.shikimoriapp.presentation.view.common.widget.NetworkErrorView
import com.gnoemes.shikimoriapp.utils.*
import com.gnoemes.shikimoriapp.utils.imageloader.ImageLoader
import com.gnoemes.shikimoriapp.utils.kotlin.gone
import com.gnoemes.shikimoriapp.utils.kotlin.visible
import com.gnoemes.shikimoriapp.utils.view.DefaultItemCallback
import com.gnoemes.shikimoriapp.utils.view.LinearStickyHead
import com.gnoemes.shikimoriapp.utils.view.ViewStatePagerAdapter
import kotlinx.android.synthetic.main.fragment_anime.view.*
import kotlinx.android.synthetic.main.fragment_comments.view.*
import kotlinx.android.synthetic.main.fragment_details_page.*
import kotlinx.android.synthetic.main.layout_default_list.view.*
import javax.inject.Inject

class AnimeFragment : BaseFragment<AnimePresenter, AnimeView>(), AnimeView {

    @Inject
    lateinit var imageLoader: ImageLoader

    @Inject
    lateinit var converter: AnimeDetailsViewModelConverter

    @Inject
    lateinit var rateResourceProvider: RateResourceProvider

    @Inject
    lateinit var userSettings: UserSettingsSource

    private val pagerAdapter: AnimePagerAdapter by lazy { AnimePagerAdapter(commentsAdapter, animeAdapter, episodeAdapter) }
    private val commentsAdapter by lazy { CommentsAdapter(imageLoader, DefaultItemCallback { presenter.onUserClicked(it) }) }
    private val episodeAdapter by lazy { EpisodeAdapter(EpisodePickCallback { item -> presenter.onEpisodeClicked(item) }, EpisodeOptionCallback(presenter::onEpisodeOptionAction)) }
    private val characterAdapter by lazy { DetailsCharacterAdapter(imageLoader, AnimeItemCallback(presenter::onAction)) }
    private val animeAdapter by lazy { AnimeAdapter(rateResourceProvider, characterAdapter, AnimeItemCallback(presenter::onAction)) }

    private var isCommentsPage: Boolean = false

    @InjectPresenter
    lateinit var animePresenter: AnimePresenter

    @ProvidePresenter
    fun providePresenter(): AnimePresenter {
        animePresenter = presenterProvider.get()
        parentFragment.ifNotNull {
            animePresenter.localRouter = (it as RouterProvider).localRouter
        }

        arguments.ifNotNull {
            presenter.setAnimeId(it.getLong(AppExtras.ARGUMENT_ANIME_ID))
        }

        return animePresenter
    }

    companion object {
        @JvmStatic
        fun newInstance(id: Long) = AnimeFragment().withArgs { putLong(AppExtras.ARGUMENT_ANIME_ID, id) }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        return inflater.inflate(fragmentLayout, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        initViews()
    }

    private fun initViews() {
        view_pager.adapter = pagerAdapter
        view_pager.addOnPageChangeListener(pageChangeListener)

        toolbar?.addBackButton()
        toolbar?.setNavigationOnClickListener { presenter.onBackPressed() }

        image_background.setOnClickListener { presenter.onBackgroundImageClicked() }
        app_bar.addOnOffsetChangedListener(onOffsetChangedListener)

        with(view_network_error) {
            setText(R.string.common_error_message_without_pull)
            gone()
        }

        presenter.onEpisodesRefresh()
    }

    private val pageChangeListener = object : ViewPager.SimpleOnPageChangeListener() {
        override fun onPageSelected(position: Int) {
            isCommentsPage = AnimeDetailsPage.COMMENTS.isEqualPage(position)
        }
    }

    private val onOffsetChangedListener = AppBarLayout.OnOffsetChangedListener { _, verticalOffset -> subtitle.alpha = (1 - Math.abs(verticalOffset.toDouble() / 1.5 / 100.0)).toFloat() }

    override fun onDestroyView() {
        view_pager.adapter = null
        view_pager.removeOnPageChangeListener(pageChangeListener)
        app_bar.removeOnOffsetChangedListener(onOffsetChangedListener)
        image_background.setOnClickListener(null)
        super.onDestroyView()
    }

    ///////////////////////////////////////////////////////////////////////////
    // GETTERS
    ///////////////////////////////////////////////////////////////////////////

    override fun getPresenter(): AnimePresenter = animePresenter

    override fun getFragmentLayout(): Int = R.layout.fragment_details_page

    ///////////////////////////////////////////////////////////////////////////
    // MVP
    ///////////////////////////////////////////////////////////////////////////

    override fun onShowLoading() {
        TransitionManager.beginDelayedTransition(coordinator_layout, Fade())
        progress_loading.visible()
        app_bar.gone()
        view_pager.gone()
    }

    override fun onHideLoading() {
        TransitionManager.beginDelayedTransition(coordinator_layout, Fade())
        progress_loading.gone()
        app_bar.visible()
        view_pager.visible()
    }

    override fun setAnimeData(details: AnimeDetailsViewModel) {
        imageLoader.setImageWithFit(image_background, details.imageUrl)
        collapsing_toolbar.title = details.title
        subtitle.text = details.subTitle
        pagerAdapter.setData(converter.convertFromViewModel(details))
    }

    override fun showEpisodeList(isEpisodeReversed: Boolean, episodes: List<BaseEpisodeItem>) {
        pagerAdapter.showEpisodeList(isEpisodeReversed, episodes)
    }

    override fun reverseEpisodes() = pagerAdapter.reverseEpisodes()
    override fun setPage(position: Int) = view_pager.setCurrentItem(position, true)
    override fun showComments(baseCommentItems: List<BaseItem>) = pagerAdapter.showComments(baseCommentItems)
    override fun insetMoreComments(baseCommentItems: List<BaseItem>) = pagerAdapter.insertMoreComments(baseCommentItems)
    override fun onShowPageLoading() = pagerAdapter.showPageLoading()
    override fun onHidePageLoading() = pagerAdapter.hidePageLoading()
    override fun onShowRefresh() = pagerAdapter.onShowRefresh()
    override fun onHideRefresh() = pagerAdapter.onHideRefresh()

    override fun showErrorView() {
        view_network_error.visible()
        view_pager.gone()
        app_bar.setExpanded(false)
    }

    override fun hideErrorView() {
        view_network_error.gone()
        view_pager.visible()
        app_bar.setExpanded(true)
    }

    override fun showPlayWizard(translationTypes: MutableList<TranslationType>?) {
        if (translationTypes != null) {
            val items = translationTypes.map { Utils.firstUpperCase(it.type) }

            if (context != null && activity?.isFinishing == false) {
                MaterialDialog(context!!.dialogContext())
                        .show {
                            listItems(items = items) { _, pos, _ ->
                                presenter.onTranslationTypeChoosed(translationTypes[pos])
                                if (userSettings.getRememberType()) {
                                    putSetting(SettingsExtras.TRANSLATION_TYPE, translationTypes[pos].type)
                                }
                            }
                            checkBoxPrompt(R.string.common_remember) { b -> putSetting(SettingsExtras.TRANSLATION_TYPE_REMEMBER, b) }
                        }
            }
        }
    }

    override fun showLinksDialog(links: List<LinkViewModel>) {
        context.ifNotNull { context ->
            MaterialDialog(context.dialogContext())
                    .show {
                        title(R.string.common_links)
                        listItems(items = links.map { it.name }) { _, pos, _ -> presenter.onLinkPressed(links[pos]) }
                    }

        }
    }

    override fun showChronologyDialog(nodes: List<FranchiseNode>) {
        context.ifNotNull { context ->
            val converter = FranchiseNodeToStringConverterImpl(context)
            MaterialDialog(context.dialogContext())
                    .show {
                        title(R.string.chronology)
                        listItems(items = converter.convertList(nodes)) { _, pos, _ ->
                            if (nodes.isNotEmpty()) {
                                presenter.onAnimeClicked(nodes[pos].id)
                            }
                        }
                    }
        }
    }

    override fun showRatesDialog(rate: UserRate?) {
        val dialog = RateDialogFragment.newInstance(Type.ANIME, rate)
        dialog.callback = object : RateDialogFragment.RateDialogCallback {
            override fun onSaveAnimeRate(rate: UserRate) = presenter.onSaveRate(rate)
            override fun onDeleteAnimeRate(id: Long) = presenter.onDeleteRate(id)
        }
        dialog.show(childFragmentManager, "RATES")
    }

    override fun showClearHistoryDialog() {
        context.ifNotNull { context ->
            MaterialDialog(context.dialogContext())
                    .show {
                        message(R.string.clear_episodes_content)
                        positiveButton(R.string.yes) { presenter.onClearHistory() }
                        negativeButton(R.string.common_cancel)
                    }
        }
    }

    ///////////////////////////////////////////////////////////////////////////
    // INNER CLASS
    ///////////////////////////////////////////////////////////////////////////

    inner class AnimePagerAdapter(
            private val commentsAdapter: CommentsAdapter,
            private val animeADapter: AnimeAdapter,
            private val episodeAdapter: EpisodeAdapter
    ) : ViewStatePagerAdapter() {

        private val screens = arrayOf(R.layout.fragment_comments, R.layout.fragment_anime, R.layout.fragment_series)

        private var commentsView: ViewGroup? = null
        private var animeView: ViewGroup? = null
        private var episodeView: ViewGroup? = null

        override fun getCount(): Int = screens.size

        override fun createView(container: ViewGroup?, position: Int): View {
            val inflater = LayoutInflater.from(container?.context)!!
            val layout = inflater.inflate(screens[position], container, false) as ViewGroup
            when (position) {
                0 -> createCommentsPage(layout)
                1 -> createAnimePage(layout)
                2 -> createSeriesPage(layout)
            }

            return layout
        }

        private fun createCommentsPage(layout: ViewGroup) {
            commentsView = layout
            commentsView.ifNotNull {
                with(it) {
                    with(list_comments) {
                        val manager = LinearLayoutManager(context)
                        itemAnimator = DefaultItemAnimator()
                        layoutManager = manager
                        adapter = commentsAdapter
                        addOnScrollListener(object : RecyclerView.OnScrollListener() {
                            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                                super.onScrolled(recyclerView, dx, dy)

                                val visibleItemPosition = manager.findLastCompletelyVisibleItemPosition() + 6
                                val itemCount = manager.itemCount - 1

                                if (visibleItemPosition >= itemCount) {
                                    presenter.loadNextPage()
                                }

                            }
                        })
                    }
                }
            }
        }

        private fun createAnimePage(layout: ViewGroup) {
            animeView = layout
            animeView.ifNotNull {
                with(it) {
                    with(anime_details_list) {
                        itemAnimator = DefaultItemAnimator()
                        layoutManager = LinearLayoutManager(context)
                        adapter = animeADapter
                    }
                }
            }
        }

        private fun createSeriesPage(layout: ViewGroup) {
            episodeView = layout
            episodeView.ifNotNull {
                with(it) {
                    refresh_layout.setOnRefreshListener { presenter.onEpisodesRefresh() }
                    this.findViewById<NetworkErrorView>(R.id.view_network_error)?.gone()
                    if (episodeAdapter.items.isEmpty()) {
                        refresh_layout.isRefreshing = true
                    }
                    with(list) {
                        itemAnimator = DefaultItemAnimator()
                        layoutManager = LinearStickyHead<EpisodeAdapter>(context)
                        addItemDecoration(DividerItemDecoration(context, LinearLayoutManager.VERTICAL))
                        adapter = episodeAdapter
                    }
                }
            }
        }

        fun setData(animeItems: List<BaseItem>) {
            animeADapter.bindItems(animeItems)
        }

        fun showEpisodeList(isEpisodeReversed: Boolean, episodes: List<BaseEpisodeItem>) {
            episodeAdapter.bindItems(isEpisodeReversed, episodes)
        }


        internal fun onHideRefresh() {
            episodeView?.refresh_layout?.isRefreshing = false
        }

        internal fun onShowRefresh() {
            episodeView?.refresh_layout?.isRefreshing = true
        }

        internal fun showComments(baseCommentItems: List<BaseItem>) {
            commentsAdapter.bindItems(baseCommentItems)
        }

        internal fun insertMoreComments(baseCommentItems: List<BaseItem>) {
            commentsAdapter.insertMore(baseCommentItems)
        }

        internal fun hidePageLoading() {
            commentsAdapter.hideProgress()
        }

        internal fun showPageLoading() {
            commentsAdapter.showProgress()
        }

        fun reverseEpisodes() {
            episodeAdapter.reverseItems()
        }

    }


}